accounts/abi/bind/backends/simulated.go

```diff
@@ -419,10 +541,38 @@ func (b *SimulatedBackend) SubscribeFilterLogs(ctx context.Context, query ethere
 	}), nil
 }
 
+// SubscribeNewHead returns an event subscription for a new header
+func (b *SimulatedBackend) SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error) {
+	// subscribe to a new head
+	sink := make(chan *types.Header)
+	sub := b.events.SubscribeNewHeads(sink)
+
+	return event.NewSubscription(func(quit <-chan struct{}) error {
+		defer sub.Unsubscribe()
+		for {
+			select {
+			case head := <-sink:
+				select {
+				case ch <- head:
```

Comment:
we have a nested select because the send to `ch` might block, and we want to be able to handle the error or quit messages while it's blocked?

It's probably simpler to avoid the nesting:

```golang
waitWrite := make(chan *types.Header)
defer close(curr)
var source <-chan *types.Header = upstream
for {
  var curr *types.Header
  select {
  case head := <-upstream:
    curr = head
  case err ...
  case quit...
 }
 select {
  case ch <- curr:
    continue
  case err...
  case quit...
  }
}
```

(unless the nesting is an established pattern in go-ethereum)

Reply:
Yeah I see nested selects everywhere:
```
node/api.go line 128
contract/oracle line 379
eth/enr_entry.go line 45
miner/miner.go line 89
p2p/simulations/network.go line 977
```

Reply:
and many more